### PR TITLE
QA kalman test

### PR DESCRIPTION
**Pull request by DonaldJennings**
## Description
Testing coverage for the Kalman Multivariate utils file. The tests are a bit short but I have tried to test everything I can see that i would have access to test.

 *Click up link* [](https://app.clickup.com/t/9gy84m)

## Changes
- Added tests for kalman_multivariate.hpp
- Added documentation to testing
- Added kalman_multivariate.test.cpp to test files

## Additional Info 
screenshots, commands to run etc.

### CODE REVIEW

File: hyped.code-workspace [hunk 1]
```diff
@@ -35,5 +35,70 @@
     "editor.detectIndentation": false,
     "C_Cpp.default.cppStandard": "c++11",
     "C_Cpp.default.cStandard": "c11",
+    "files.associations": {
```
Comment (mifrandir) [inline hyped.code-workspace]: Why is this here? What does this do? Has this got anything to do with the Kalman tests?

File: test/src/utils/kalman_multivariate.test.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  MatrixXf R_Data[100];
+  MatrixXf P_Data[100];
+  MatrixXf I = MatrixXf::Identity(n, n);
+
+  void SetUp()
+  {
+    // Populates the arrays defined above with 100 random values
+    for (int i = 0; i < 100; i++) {
+      VectorXf x1 = VectorXf::Random(n);
+      x1_Data[i] = x1;
```
Comment (mifrandir) [inline test/src/utils/kalman_multivariate.test.cpp]: Why are you not using the following?

```cpp
x1_Data[i] = VectorXf::Random(n);
```

It seems a bit bloated this way. Unless there is a good reason or something in the style guide that I am missing.

File: test/src/utils/kalman_multivariate.test.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+#include "utils/system.hpp"
+
+using hyped::utils::math::KalmanMultivariate;
+
+// -------------------------------------------------------------------------------------------------
+// Functionality
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * Struct used foor test fixtures testing the functionality of Kalman_Multivaraite.hpp. Constructor
```
Comment (mifrandir) [inline test/src/utils/kalman_multivariate.test.cpp]: Typo...

File: test/src/utils/kalman_multivariate.test.cpp (hunk 2 above)
Comment (DonaldJennings) [reply test/src/utils/kalman_multivariate.test.cpp]: Yeah thats a much better shout.
Thanks!

File: test/src/utils/kalman_multivariate.test.cpp [hunk 4]
```diff
@@ -67,7 +67,7 @@ struct KalmanFunctionality : public ::testing::Test {
  */
 TEST_F(KalmanFunctionality, handlesZeroStateEstimate)
 {
-  ASSERT_EQ(kalman.getStateEstimate(), x0);
+  ASSERT_EQ(kalman.getStateEstimate(), x0) << "Should handle zero vector as state estimate";
```
Comment (pmorande27) [inline test/src/utils/kalman_multivariate.test.cpp]: Maybe it would be a good Idea to encapsulate the error messages in variables. I think it is cleaner that way

File: hyped.code-workspace (hunk 1 above)
Comment (pmorande27) [reply hyped.code-workspace]: Yep this is weird, I don't think that you should change this file.

File: test/src/Test.files [hunk 5]
```diff
@@ -6,3 +6,4 @@ T_SRCS := \
   state_machine/dummy.test.cpp \
   utils/config.test.cpp \
   utils/differentiator.test.cpp \
+  utils/kalman_multivariate.test.cpp \
```
Comment (pmorande27) [inline test/src/Test.files:8]: You are going to have a merge conflict with this file, but it should be easy to solve.

File: test/src/utils/kalman_multivariate.test.cpp (hunk 4 above)
Comment (DonaldJennings) [reply test/src/utils/kalman_multivariate.test.cpp]: Yeah I agree. We need to have a discussion about this when Sem 2 starts to properly define the style. Perhaps creating a simple variable in the Test Struct storing the message

File: test/src/utils/kalman_multivariate.test.cpp [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+ */
+struct KalmanMathematics : public::testing::Test {
+  unsigned int n = 3;
+  unsigned int m = 2;
+  unsigned int k = 1;
+
+ protected:
+  KalmanMultivariate kalmanMathWithoutControl = KalmanMultivariate(n, m, 0);
+  KalmanMultivariate kalmanMathWithControl = KalmanMultivariate(n, m, k);
+  VectorXf x1_Data[100];
```
Comment (mifrandir) [inline test/src/utils/kalman_multivariate.test.cpp]: I'm not sure about the unnamed constant `100`. I know very little about Kalman filters and it's not obvious to me whether this is an arbitrary testing value or essential for this code to work.

**Review by flori-margaritescu — APPROVED** [review]

**Review by mifrandir — APPROVED** [review]
I do not know anything about Kalman filters. That being said, these tests look idiomatic and well written to me.

**Review by pmorande27 — APPROVED** [review]
Everything Ok now!
